### PR TITLE
Use tiered projections for hash aggregates

### LINKED ISSUE #5085: Could we "push down" some of projections that are evaluated by `GpuExtractChunk32`
The `GpuExtractChunk32` expression gets evaluated 4 times for a decimal column (as far as I understand). This expression could also have some child expressions, for example:

`gpudecimal128sum((cast(ss_quantity#139 as decimal(10,0)) * ss_sales_price#142)`

So in this case, the multiply and the cast are both required to be re-evaluated 4 times. This means that for the aggregate we could spend most of the time in the pre-processing steps (I see this in queries q4, q5, q14*, q23*, q67). See the red `GpuExtractChunk32` blocks, and the libcudf calls under it. They can be non-trivial amounts of time:

Screenshot from 2022-03-29 13-43-31

This issue is to track coming up with a way to extract this child expression and evaluate it once if possible.

### PR DESCRIPTION
**Pull request by jbrennan333**
Signed-off-by: Jim Brennan <jimb@nvidia.com>

Closes #5085.

This implements a tiered approach for input projections in hash aagregations.  This is a performance optimization to reduce computation by pulling out common sub-expressions and doing them once.  The general approach is to provide an alternative to bindGpuReferences, currently called bindGpuReferencesTiered, which returns a case class that contains a tiered list of references, and the corresponding method in that case class that applies the tiered projections in order. The tiered list of references are created by recursively pulling out common sub-expressions from the initial list of expressions and rewriting expressions to refer to the output of the sub-expressions.

This feature is enabled by default, but there is a configuration property, `spark.rapids.sql.tiered.project.enabled` that can be used to disable it.

See #5085 for more discussion and performance.  If this works well for aggregations, we will consider enabling it for other projections in the plugin.

### CODE REVIEW

**Comment by jbrennan333** [comment]
build

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/catalyst/expressions/GpuEquivalenExpressions.scala
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/catalyst/expressions/GpuEquivalenExpressions.scala:257]: Are these needed still?

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/catalyst/expressions/GpuEquivalenExpressions.scala
Comment (revans2) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/catalyst/expressions/GpuEquivalenExpressions.scala:283]: Didn't we replace EquivalentExpressions???

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/catalyst/expressions/GpuEquivalenExpressions.scala
Comment (abellina) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/catalyst/expressions/GpuEquivalenExpressions.scala:214]: if this is only used in testing, it may be nice to mark it as so, like `getAllExprStates` (nit). Could it also be `private[sql]`?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala
Comment (abellina) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala:260]: it would be helpful, if there could be minimal example in this header that shows the projected tier process for a couple of tiers and a simple original `cb`. It took me a while to grok how this function works.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala
Comment (abellina) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala:723]: I wish, and it could be follow on work, that this could be handled by `GpuProjectExec.project`. E.g. that `preStepBoundTiered` could be a class, something like `TieredBoundReferences`,  that `.project` could be overloaded for. It would remove some of the changes here, and push the decision to use tiered to the binding part.

Comment (jbrennan333) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala:723]: I agree that would be better.  I may defer this to a follow-on PR though.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala
Comment (jbrennan333) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala:260]: I will add a comment.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala
Comment (jbrennan333) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala:264]: Yes.  I should be able to make this change.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/catalyst/expressions/GpuEquivalenExpressions.scala
Comment (jbrennan333) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/catalyst/expressions/GpuEquivalenExpressions.scala:77]: This method and comment are from the spark EquivalentExpressions code.  The spark code effectively only gives you one tier of substitutions.  To get multiple tiers, I am calling the spark code recursively in getExprTiers().
In your example, I think the spark code will give you:
```
((a + b) + c), (a + b)
```
I then recurse on this to create another tier, so we end up with:
```
(a+b) as tmp1
(tmp1 + c) as tmp2, tmp1 as tmp3,
tmp2, (tmp2 + d), tmp3
```

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/catalyst/expressions/GpuEquivalenExpressions.scala
Comment (jbrennan333) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/catalyst/expressions/GpuEquivalenExpressions.scala:214]: Yes.  I will fix this.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/catalyst/expressions/GpuEquivalenExpressions.scala
Comment (jbrennan333) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/catalyst/expressions/GpuEquivalenExpressions.scala:257]: I am still using them for debugging, but I will remove before merging.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/catalyst/expressions/GpuEquivalenExpressions.scala
Comment (jbrennan333) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/catalyst/expressions/GpuEquivalenExpressions.scala:283]: Yes. And I removed that filter as well.  I will remove the comment.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala
Comment (jbrennan333) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala:276]: Agreed.  Removing eclipsed columns would be a good improvement.  It's a little complicated, because the original set of expressions could still have references to a and b that are separate from instances of (a + b), like (a + c) or (a +d).  But we should be able to determine this.   Just adding the subexpression result columns was the easier first step.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala
Comment (jbrennan333) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala:805]: I will remove this.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala
Comment (revans2) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala:276]: If you have the tiers before you bind the expressions, then it is not that bad.

Look at the inputs (leaf nodes) of the output tier, this is what the output of the previous tier needs to be. If there is something missing from that tier, then add it in as a separate column that just references that column.  Then go to the next tier and look again.  Repeat until all of the tiers are done.

After that you can bind each tier the regular way to the outputs of the previous tier...

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/catalyst/expressions/GpuEquivalenExpressions.scala
Comment (revans2) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/catalyst/expressions/GpuEquivalenExpressions.scala:77]: okay.

**Comment by jbrennan333** [comment]
I pushed up a commit that addresses some of the review comments.  I am still doing some additional rework, so please hold off on merging for now.

**Comment by jbrennan333** [comment]
build

**Comment by jbrennan333** [comment]
Updated to address more review comments.  I've done a significant rewrite of getExprTiers to make it more efficient and to fix a problem I was seeing where it was making aliases of aliases in each tier, increasing the number of columns (at least column references) pretty significantly.  It no longer does this.  A simple example is from an earlier comment, these expressions: `((a + b) + c), (a + b)`  produced this set of expression tiers:
```
(a+b) as tmp1
(tmp1 + c) as tmp2, tmp1 as tmp3,
tmp2, (tmp2 + d), tmp3
```
Now this will produce:
```
(a+b) as tmp1
(tmp1 + c) as tmp2
tmp2, (tmp2 + d), tmp1
```
The redundant `tmp1 as tmp3` is removed.  It's small in this example, but in some queries, like q66, there can be hundreds of these.  I don't think it was causing much extra work, because they are all aliases, but it's much cleaner now.

This update does not include removing eclipsed columns.  I will address that in a follow-up PR.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala
Comment (jbrennan333) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala:269]: Comment added.  I also added a unit test like what the comment says.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/catalyst/expressions/GpuEquivalenExpressions.scala
Comment (jbrennan333) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/catalyst/expressions/GpuEquivalenExpressions.scala]: Removed.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala
Comment (jbrennan333) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala]: Removed.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala
Comment (jbrennan333) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala]: Thanks for the suggestion.  I rewrote it as tail recursive.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/catalyst/expressions/GpuEquivalenExpressions.scala
Comment (jbrennan333) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/catalyst/expressions/GpuEquivalenExpressions.scala]: Comment removed.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/catalyst/expressions/GpuEquivalenExpressions.scala
Comment (jbrennan333) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/catalyst/expressions/GpuEquivalenExpressions.scala]: done.

**Comment by jbrennan333** [comment]
build

**Comment by jbrennan333** [comment]
I added some debugging to print out number of columns added and batch sizes for each tier when tiered project is happening.  I ran nds2.0 power run on spark2a and captured this data.  I split it out into three files, one for 2 tiers (1 sub-expression tier), one for 3 tiers (2 sub-expression tiers), and one for 4 tiers (3 sub-expression tiers).  In each of these files, each line looks something like this:
```
cols(in,diff,diff,out)  5  2  2 13      sizes(in,diff,diff,out)           5,044           3,984           7,968          13,510
```
This line shows 5 columns in the input batch, 2 columns added in the first tier, 2 columns added in the second tier, and 13 columns in the output batch.

It also shows a batch size of 5,044 for the input batch and 13,510 for the output batch.
The other batch sizes are differentials from the previous column, so the first tiered project added 3,984 bytes, and the second tier added 7,968 bytes.

The files are sorted by the last differential batch size column.

[batch-data-2-tiers-spark2a.txt](https://github.com/NVIDIA/spark-rapids/files/9508334/batch-data-2-tiers-spark2a.txt)
[batch-data-3-tiers-spark2a.txt](https://github.com/NVIDIA/spark-rapids/files/9508338/batch-data-3-tiers-spark2a.txt)
[batch-data-4-tiers-spark2a.txt](https://github.com/NVIDIA/spark-rapids/files/9508340/batch-data-4-tiers-spark2a.txt)

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala
Comment (abellina) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala:269]: Thanks!

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala
Comment (revans2) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala]: We will do this as follow on work so I am fine with putting this in as is now.

**Review by revans2 — APPROVED** [review]

**Comment by jbrennan333** [comment]
Thanks for the reviews @revans2 and @abellina!

**Pull request merged**
